### PR TITLE
Feat/swift: missing events

### LINKED ISSUE #1542: [Task]: Update lib.rs based on nodejs binding for Swift
### Task description

We have them in the nodejs bindings and we would need also have them in the swift bindings

### Requirements

TBD

### Acceptance criteria

TBD

### Creation checklist

- [X] I have assigned this task to the correct people
- [X] I have added the most appropriate labels
- [ ] I have linked the correct milestone and/or project

### PR DESCRIPTION
**Pull request by kwek20**
# Description of change

Added the 2 missing methods for using wallet with capacitor in OSX

## Links to any relevant issues

fixes issue #1542

## Type of change
- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

- [x] Ran on OSX

### CODE REVIEW

File: bindings/swift/Cargo.toml [hunk 1]
```diff
@@ -9,8 +9,10 @@ edition = "2021"
 crate-type = ["cdylib", "staticlib"]      # Creates dynamic lib
 
 [dependencies]
+fern-logger = { version = "0.5.0", default-features = false }
 futures =  { version = "0.3.23", default-features = false }
-iota-wallet = { path = "../../", version = "0.2.0", features = ["message_interface", "events"] }
+iota-wallet = { path = "../../", features = ["message_interface", "events"] }
```
Comment (thibault-martinez) [inline bindings/swift/Cargo.toml]:
```suggestion
iota-wallet = { path = "../../", features = [ "message_interface", "events" ] }
```

Comment (kwek20) [reply bindings/swift/Cargo.toml]: we should add features `"mnemonic", "storage", "stronghold"` to be consistent with how it was

Comment (thibault-martinez) [reply bindings/swift/Cargo.toml]: My bad, I edited. I just really wanted the spaces :)

File: bindings/swift/Cargo.toml [hunk 2]
```diff
@@ -6,12 +6,14 @@ edition = "2021"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [lib]
-crate-type = ["cdylib", "staticlib"]      # Creates dynamic lib
+crate-type = [ "cdylib", "staticlib" ]      # Creates dynamic lib
 
 [dependencies]
-iota-wallet = { path = "../../", version = "0.2.0", features = [ "message_interface", "events" ] }
+iota-wallet = { path = "../../",  features = [ "message_interface", "events" ] }
```
Comment (thibault-martinez) [inline bindings/swift/Cargo.toml]:
```suggestion
iota-wallet = { path = "../../", features = [ "message_interface", "events" ] }
```

File: bindings/swift/xcode/IotaWallet/IotaWallet/Wallet.m [hunk 3]
```diff
@@ -15,6 +15,9 @@ @interface Wallet () {
 
 @implementation Wallet
 
++ (BOOL) init_logger:(NSString*) file_name level_filter: (nullable NSString*) {
```
Comment (amadeu2) [inline bindings/swift/xcode/IotaWallet/IotaWallet/Wallet.m]: Seems this is lacking when opened with xcode
```suggestion
+ (BOOL) init_logger:(NSString*) file_name level_filter: (nullable NSString*) level_filter {
```

**Review by amadeu2 — COMMENTED** [review]
To complete my tests / review I would need to update my macOS VM cause seems needs a new version of XCode. Even though this is only needed for the Swift package, but not for the iOS Capacitor plugin where is only needed the C Header `iota_wallet_ffi.h`

File: bindings/swift/Cargo.toml [hunk 4]
```diff
@@ -15,4 +15,7 @@ futures =  { version = "0.3.25", default-features = false }
 once_cell = { version = "1.16.0", default-features = false }
 serde = { version = "1.0.147", default-features = false, features = [ "derive" ] }
 serde_json = { version = "1.0.88", default-features = false }
-tokio = { version = "1.22.0", default-features = false}
+tokio = { version = "1.22.0", default-features = false }
+
```
Comment (thibault-martinez) [inline bindings/swift/Cargo.toml]: Remove the new line and order them within the rest

Comment (kwek20) [reply bindings/swift/Cargo.toml]: Sad times, i thought logging separate :)

File: bindings/swift/iota_wallet_ffi.h [hunk 5]
```diff
@@ -13,10 +13,14 @@ typedef struct iota_wallet_handle iota_wallet_handle_t;
 
 typedef void (*Callback)(const char* response, const char* error, void* context);
 
+// Initialise the fern logger. Level filter defaults to Debug if omitted
+extern int8_t iota_init_logger(const char* file_name, const char* level_filter);
 extern iota_wallet_handle_t* iota_initialize(const char* manager_options, char* error_buffer, size_t error_buffer_size);
 extern void iota_destroy(iota_wallet_handle_t*);
 extern void iota_send_message(iota_wallet_handle_t* wallet_handle, const char* message, Callback callback, void* context);
 extern int8_t iota_listen(iota_wallet_handle_t* wallet_handle, const char* event_types, Callback callback, void* context, char* error_buffer, size_t error_buffer_size);
+extern int8_t iota_clear_listeners(iota_wallet_handle_t* wallet_handle, const char* event_types, Callback callback, void* context, char* error_buffer, size_t error_buffer_size);
```
Comment (thibault-martinez) [inline bindings/swift/iota_wallet_ffi.h:22]: I'm not a very big fan of the prefixes. We don't have that in the other bindings do we ?

File: bindings/swift/src/lib.rs [hunk 6]
```diff
@@ -208,16 +210,76 @@ pub unsafe extern "C" fn iota_listen(
...
-//     let output_config = LoggerOutputConfigBuilder::new()
-//         .name(file_name)
-//         .level_filter(LevelFilter::Debug);
-//     init_logger(LoggerConfig::build().with_output(output_config));
-// }
+/// # Safety
+///
+/// `callback` will be invoked from another thread so context must be safe to `Send` across threads
+#[no_mangle]
+pub unsafe extern "C" fn clear_listeners(
```
Comment (thibault-martinez) [inline bindings/swift/src/lib.rs]: How does this name relate to the iota_ definitions ? Shouldn't this have an iota_prefix as well?

**Review by amadeu2 — DISMISSED** [review]
The needed the C Header `iota_wallet_ffi.h` for the Capacitor plugin is working for me on iOS phones and simulators. There are no errors now on XCode either with the Swift package, thanks! 💯

File: bindings/swift/iota_wallet_ffi.h (hunk 5 above)
Comment (kwek20) [reply bindings/swift/iota_wallet_ffi.h:22]: No, but this is a C thing, as there are no namespaces, this is how most people make their methods

File: bindings/swift/src/lib.rs (hunk 6 above)
Comment (kwek20) [reply bindings/swift/src/lib.rs]: Good catch! Thats not as intended and doesnt work, ive added the prefix

File: bindings/swift/iota_wallet_ffi.h (hunk 5 above)
Comment (thibault-martinez) [reply bindings/swift/iota_wallet_ffi.h:22]: Fair enough

File: bindings/swift/src/lib.rs (hunk 6 above)
Comment (thibault-martinez) [reply bindings/swift/src/lib.rs]: Added an issue for swift bindings CI to catch that kind of stuff

File: bindings/swift/src/lib.rs [hunk 7]
```diff
@@ -208,16 +210,76 @@ pub unsafe extern "C" fn iota_listen(
...
-//         CStr::from_ptr(file_name)
-//     };
-//     let _file_name = c_file_name.to_str().unwrap();
-
-//     let output_config = LoggerOutputConfigBuilder::new()
-//         .name(file_name)
-//         .level_filter(LevelFilter::Debug);
-//     init_logger(LoggerConfig::build().with_output(output_config));
-// }
+/// # Safety
```
Comment (thibault-martinez) [inline bindings/swift/src/lib.rs:213]:
```suggestion

/// # Safety
```

**Review by Thoralf-M — APPROVED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**